sample/trace_key.go

```diff
@@ -40,75 +43,194 @@ func newTraceKey(fields []string, useTraceLength bool) *traceKey {
 					break outer
 				}
-				fieldCollector[field] = append(fieldCollector[field], fmt.Sprintf("%v", val))
+				if uniques.AddAsString(val, i) {
+					continue
+				}
 			}
 		}
 	}
 	// ok, now we have a map of fields to a list of all unique values for that field.
 	// (unless it was huge, in which case we have a bunch of them)
 
-	var key strings.Builder
-	for _, field := range d.fields {
+	// please change this so that strings.Builder can be reused between calls
+
+	for i := range d.fields {
+		values := uniques.Values(i)
 		// if there's no values for this field, skip it
-		values, ok := fieldCollector[field]
-		if !ok || len(values) == 0 {
+		if len(values) == 0 {
 			continue
 		}
-		// sort and collapse list
-		sort.Strings(values)
 		var prevStr string
 		for _, str := range values {
 			if str != prevStr {
-				key.WriteString(str)
-				key.WriteRune('•')
+				d.keyBuilder.WriteString(str)
+				d.keyBuilder.WriteRune('•')
 				fieldCount += 1
 			}
 			prevStr = str
 		}
 		// get ready for the next element
-		key.WriteRune(',')
+		d.keyBuilder.WriteRune(',')
 	}
 
 	if trace.RootSpan != nil {
 		for _, field := range d.rootOnlyFields {
 			if val, ok := trace.RootSpan.Data[field]; ok {
-				key.WriteString(fmt.Sprintf("%v,", val))
+				d.keyBuilder.WriteString(fmt.Sprintf("%v,", val))
 				fieldCount += 1
 			}
 		}
 	}
 
 	if d.useTraceLength {
-		key.WriteString(strconv.FormatInt(int64(len(spans)), 10))
+		d.keyBuilder.WriteString(strconv.FormatInt(int64(len(spans)), 10))
 		fieldCount += 1
 	}
 
-	return key.String(), fieldCount
+	return d.keyBuilder.String(), fieldCount
+}
+
+// distinctValue keeps track of distinct values for a set of fields.
+// It stores the unique values as strings.
+type distinctValue struct {
+	buf          []byte
+	fields       []string
+	values       []map[uint64]string
+	valuesBuffer []string
+
+	// totalUniqueCount keeps track of how many unique values we've seen so far for a trace key.
+	totalUniqueCount int
+	// maxDistinctValue is the maximum number of distinct values we will store for a trace key.
+	maxDistinctValue int
+}
+
+func (d *distinctValue) init(fields []string, maxDistinctValue int) {
+	for i := range fields {
+		if i >= len(d.fields) {
+			// if we don't have enough fields, allocate more
+			d.fields = append(d.fields, fields[i])
+			continue
+		}
+		d.fields[i] = fields[i]
+	}
+
+	if len(d.values) < len(fields) {
+		// if we don't have enough values, allocate more
+		for i := len(d.values); i < len(fields); i++ {
+			d.values = append(d.values, make(map[uint64]string))
+		}
+	} else {
+		// if we have more values than fields, trim the excess
+		d.values = d.values[:len(fields)]
+	}
+
+	d.maxDistinctValue = maxDistinctValue
+}
+
+func (d *distinctValue) Reset() {
+	// Reset the fields and values but do not reallocate them
+	for i := range d.fields {
+		d.fields[i] = ""
+	}
+	for i := range d.values {
+		if d.values[i] != nil {
+			for k := range d.values[i] {
+				delete(d.values[i], k)
```

Comment:
As I noted about you should use `clear()` - deleting entries from a map one by one is very slow.